### PR TITLE
upload csv to database and display as table

### PR DESCRIPTION
**Pull request by t-kanin**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -20,5 +20,6 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
 
+  has_many :keywords, dependent: :destroy
```
Comment (aunsira) [inline app/models/user.rb:23]: Let's add a line break.

File: app/views/keywords/_keyword.html.erb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<tr>
+  <td><%= keyword.keyword%></td>
```
Comment (aunsira) [inline app/views/keywords/_keyword.html.erb]:
```suggestion
  <td><%= keyword.keyword %></td>
```

File: app/views/keywords/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      <th>Keyword</th>
+      <th># of Adwords</th>
+      <th># of Non-Adwords</th>
+      <th># of links</th>
+      <th># of search links</th>
+    </tr>
+  </thead>
+  <tbody id="myTable">
+    <tr>
+      <% if @keywords.any?%>
```
Comment (aunsira) [inline app/views/keywords/index.html.erb]:
```suggestion
      <% if @keywords.any? %>
```

File: app/views/keywords/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        <%= render @keywords %>
+      <%end%>
```
Comment (aunsira) [inline app/views/keywords/index.html.erb]:
```suggestion
      <% end %>
```

File: app/models/keyword.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+class Keyword < ApplicationRecord
+  belongs_to :user
+
+  def self.import(file, uid)
+    options = {}
+    SmarterCSV.process(file.path, options) do |chunk|
+      chunk.each do |data_hash|
+        data_hash[:user_id] = uid
+        Keyword.create!(data_hash)
```
Comment (aunsira) [inline app/models/keyword.rb]: Improvement: Consider using `insert_all`.

https://api.rubyonrails.org/classes/ActiveRecord/Persistence/ClassMethods.html#method-i-insert_all

File: app/models/keyword.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class Keyword < ApplicationRecord
+  belongs_to :user
+
+  def self.import(file, uid)
+    options = {}
+    SmarterCSV.process(file.path, options) do |chunk|
```
Comment (aunsira) [inline app/models/keyword.rb:12]: Should we validate the csv file first before start importing? For example file is not empty / the content type of file is really csv /  the upload size of keyword.

File: config/application.rb [hunk 7]
```diff
@@ -1,6 +1,7 @@
 require_relative "boot"
 
 require "rails"
+require 'smarter_csv'
```
Comment (aunsira) [inline config/application.rb]: Not sure but we might not need to do this by ourself.

File: app/controllers/keywords_controller.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class KeywordsController < ApplicationController
+  def index
+    @keywords = Keyword.all
```
Comment (WatcharitNew) [inline app/controllers/keywords_controller.rb]: Not sure that it has to show only current user or not

File: spec/models/keyword_spec.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Keyword, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (WatcharitNew) [inline spec/models/keyword_spec.rb:4]: May add test association with user

File: db/migrate/20210916032614_add_default_time_stamp_to_keyword.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class AddDefaultTimeStampToKeyword < ActiveRecord::Migration[6.1]
+  def change
+    change_column_default :keywords, :created_at, -> { 'CURRENT_TIMESTAMP' }
+    change_column_default :keywords, :updated_at, -> { 'CURRENT_TIMESTAMP' }
```
Comment (WatcharitNew) [inline db/migrate/20210916032614_add_default_time_stamp_to_keyword.rb:4]: I think that default timestamp is already current timestamp

Comment (t-kanin) [reply db/migrate/20210916032614_add_default_time_stamp_to_keyword.rb:4]: The problem arises when you do insert_all function with not null validation error. https://github.com/rails/rails/issues/35493

File: app/services/file_handler.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  class FileEmptyError < StandardError
+    def message
+      'Please select file to upload'
+    end
+  end
+
+  def self.handle_upload(file, uid)
+    raise FileEmptyError if file.nil?
+    raise FileExntensionError unless %w[.csv].include? File.extname(file)
+    raise FileTooBigError if file.size > 1000.megabytes
```
Comment (aunsira) [inline app/services/file_handler.rb]: The assignment says we should validate the number of keywords, not the file size, isn't it?

**Review by aunsira — APPROVED** [review]
Let's fix things in next PR.

**Pull request merged**